src/libmongoc/src/mongoc/mongoc-server-monitor.c

```diff
@@ -0,0 +1,1241 @@
+                      "Malformed message from server");
+      GOTO (fail);
+   }
+
+   if (!_mongoc_rpc_decompress_if_necessary (&rpc, &buffer, error)) {
+      GOTO (fail);
+   }
+
+   _mongoc_rpc_swab_from_le (&rpc);
+   memcpy (&msg_len, rpc.msg.sections[0].payload.bson_document, 4);
+   msg_len = BSON_UINT32_FROM_LE (msg_len);
+   if (!bson_init_static (
+          &reply_local, rpc.msg.sections[0].payload.bson_document, msg_len)) {
+      bson_set_error (error,
+                      MONGOC_ERROR_PROTOCOL,
+                      MONGOC_ERROR_PROTOCOL_INVALID_REPLY,
+                      "Malformed BSON payload from server");
+      GOTO (fail);
+   }
+
+   bson_copy_to (&reply_local, ismaster_reply);
+   server_monitor->more_to_come =
+      (rpc.msg.flags & MONGOC_MSG_MORE_TO_COME) != 0;
+
+   ret = true;
+fail:
+   if (!ret) {
+      bson_init (ismaster_reply);
+   }
+   _mongoc_buffer_destroy (&buffer);
+   return ret;
+}
+
+/* Send and receive an awaitable ismaster.
+ *
+ * Called only from server monitor thread.
+ * Locks server monitor mutex.
+ * May block for up to heartbeatFrequencyMS waiting for reply.
+ */
+static bool
+_server_monitor_awaitable_ismaster (mongoc_server_monitor_t *server_monitor,
+                                    const bson_t *topology_version,
+                                    bson_t *ismaster_reply,
+                                    bool *cancelled,
+                                    bson_error_t *error)
+{
+   bson_t cmd = BSON_INITIALIZER;
+   bool ret = false;
+
+   bson_append_int32 (&cmd, "isMaster", 8, 1);
+   _server_monitor_append_cluster_time (server_monitor, &cmd);
+   bson_append_document (&cmd, "topologyVersion", 15, topology_version);
+   bson_append_int32 (
+      &cmd, "maxAwaitTimeMS", 14, server_monitor->heartbeat_frequency_ms);
+   bson_append_utf8 (&cmd, "$db", 3, "admin", 5);
+
+   if (!_server_monitor_awaitable_ismaster_send (server_monitor, &cmd, error)) {
+      GOTO (fail);
+   }
+
+   if (!_server_monitor_awaitable_ismaster_recv (
+          server_monitor, ismaster_reply, cancelled, error)) {
+      bson_destroy (ismaster_reply);
+      GOTO (fail);
+   }
+
+   ret = true;
+fail:
+   if (!ret) {
+      bson_init (ismaster_reply);
+   }
+   bson_destroy (&cmd);
+   return ret;
+}
+
+/* Update the topology description with a reply or an error.
+ *
+ * Called only from server monitor thread.
+ * Caller must hold no locks.
+ * Locks topology mutex and server monitor mutex.
+ */
+static void
+_server_monitor_update_topology_description (
+   mongoc_server_monitor_t *server_monitor,
+   mongoc_server_description_t *description)
+{
+   mongoc_topology_t *topology;
+   bson_t *ismaster_reply = NULL;
+
+   topology = server_monitor->topology;
+   if (description->has_is_master) {
+      ismaster_reply = &description->last_is_master;
+   }
+
+   if (ismaster_reply) {
+      _mongoc_topology_update_cluster_time (topology, ismaster_reply);
```

Comment:
`_mongoc_topology_update_cluster_time` also takes the lock internally, but I suppose there's no way around that unless you were to extract the update logic into a separate function so it could be called while the lock is already held -- allowing you to consolidate with the lock you take out below.

Probably nothing to do here, but I wanted to point it out in case there's a concern with taking out locks too frequently.